test/viewer.js

```diff
@@ -0,0 +1,50 @@
+const net = require('net');
+
+const Logger = require('../lib/Logger');
+const { startServer } = require('../lib/viewer.js');
+
+describe('WebSocket server', function () {
+  it('should not crash when an error is emitted on the websocket', function (done) {
+    const bundleStats = {
+      assets: [{ name: 'bundle.js', chunks: [0] }]
+    };
+
+    const options = {
+      openBrowser: false,
+      logger: new Logger('silent'),
+      port: 0
+    };
+
+    startServer(bundleStats, options).then(function ({ http: server }) {
+      const socket = net.createConnection(server.address().port, function () {
+        socket.write([
+          'GET / HTTP/1.1',
+          'Host: localhost',
+          'Upgrade: websocket',
+          'Connection: Upgrade',
+          'Sec-WebSocket-Key: Ix02x1E8kmC9ie01/dGCuw==',
+          'Sec-WebSocket-Version: 13',
+          '\r\n'
+        ].join('\r\n'));
+      });
+
+      socket.on('data', function (chunk) {
+        const expected = Buffer.from([
+          'HTTP/1.1 101 Switching Protocols',
+          'Upgrade: websocket',
+          'Connection: Upgrade',
+          'Sec-WebSocket-Accept: XYMa8DJmsxHNEsGJZwEZT7XX1Dk=',
```

Comment:
Could we generate this response using the `crypto` module instead of hard-coding it? I didn't know about these headers, and it would've been nice to know where this magical looking value came from.

```js
const crypto = require('crypto');
const websocketAcceptHash = crypto.createHash('sha1');
websocketAcceptHash.update(
  // The client-generated "Sec-WebSocket-Key" value
  'Ix02x1E8kmC9ie01/dGCuw==' +
  // The GUID constant defined in WebSocket protocol draft
  // https://tools.ietf.org/html/draft-ietf-hybi-thewebsocketprotocol-17#section-1.3
  '258EAFA5-E914-47DA-95CA-C5AB0DC85B11'
);
const websocketAccept = hash.digest('base64');
```

Reply:
Sure.